### PR TITLE
feat(multi actions): execution loop v0

### PR DESCRIPTION
**Pull request by fontanierh**
## Description

https://github.com/dust-tt/dust/issues/4602

This commit adds a first version of the multi-actions execution loop.
`generation` is not yet an action, so it is currently treated differently from other actions. generation is always the last thing done in the loop (and currently, is always performed). This should be changed once generation is an action.

Right now, the `generation` is also not counted towards the maxToolsUsePerRun, which may not be what we want

## Risk

Should not impact single action assistants

### CODE REVIEW

**Review by flvndvd — APPROVED** [review]
LGTM, left a few suggestions

File: front/lib/api/assistant/agent.ts [hunk 1]
```diff
@@ -80,9 +97,149 @@ export async function* runAgent(
...
+      });
+      for await (const event of eventStream) {
+        yield event;
+      }
+    } else {
+      // TODO(@fontanierh): remove this once generation is part of actions.
+      // This is just to assert that we cover all action types.
+      ((g: AgentGenerationConfigurationType) => {
+        void g;
+      })(action);
```
[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:174]: The self-executing function is not meant to break the loop, but rather to make sure every action type is handled (we'd get a type error if `action`  was not a generation config

File: front/lib/api/assistant/agent.ts [hunk 2]
```diff
@@ -80,9 +97,149 @@ export async function* runAgent(
...
+      conversationId: conversation.sId,
+      multiActionLoopIteration: i,
+    });
+
+    localLogger.info("Starting multi-action loop iteration");
+
+    const forcedAction = fullConfiguration.actions.find(
+      (a) => a.forceUseAtIteration === i
+    );
+    const actions = forcedAction ? [forcedAction] : fullConfiguration.actions;
```
[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:116]: indeed 👍

[the remark this answers is not in the document]
Comment (spolu) [reply front/lib/api/assistant/agent.ts:116]: We could:
(i) set `function_call` and keep all specifications
(ii) set `function_call` and send only one specification
(iii) [current] send only one specification

(ii) > (iii) so we def want it.
(i) is a bit more principled than (ii) + easier to code but consumes a bit more context tokens.

I would therefore got for (ii)

**Review by spolu — CHANGES_REQUESTED** [review]
Looks great. Left a small change request

File: front/lib/api/assistant/agent.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:116]: (ii) seems good indeed 👍  on it

File: front/lib/api/assistant/agent.ts [hunk 3]
```diff
@@ -80,9 +97,149 @@ export async function* runAgent(
...
+      // TODO(@fontanierh): remove this once generation is part of actions.
+      // This is just to assert that we cover all action types.
+      ((g: AgentGenerationConfigurationType) => {
+        void g;
+      })(action);
+      break;
+    }
   }
+
+  const eventStream = runGeneration(
```
[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts]: Really I don't know how to call it (we already have a "runGeneration" function). Keeping as-is for now

File: front/lib/api/assistant/agent.ts [hunk 4]
```diff
@@ -270,14 +310,414 @@ export async function getNextAction(
...
+      });
+      for await (const event of eventStream) {
+        yield event;
+      }
+    } else {
+      // TODO(@fontanierh): remove this once generation is part of actions.
+      // This is just to assert that we cover all action types.
+      ((g: AgentGenerationConfigurationType) => {
+        void g;
+      })(action);
```
Comment (flvndvd) [inline front/lib/api/assistant/agent.ts:417]: Why do we need those lines of code for?

File: front/lib/api/assistant/agent.ts [hunk 5]
```diff
@@ -463,6 +326,167 @@ export async function getNextAction(
   });
 }
 
+export async function* runMultiActionsAgent(
```
Comment (flvndvd) [inline front/lib/api/assistant/agent.ts:329]: ❤️

File: front/lib/api/assistant/agent.ts [hunk 6]
```diff
@@ -198,19 +232,25 @@ export async function getNextAction(
...
+    });
+  }
 
   const config = cloneBaseConfig(
     DustProdActionRegistry["assistant-v2-use-tools"].config
   );
   config.MODEL.function_call = "auto";
   config.MODEL.provider_id = model.providerId;
   config.MODEL.model_id = model.modelId;
+  if (specifications.length === 1) {
```
Comment (flvndvd) [inline front/lib/api/assistant/agent.ts:251]: 🧐 I'm curious if we should only apply this change for the `forceUseAtIteration` cases? Imagine a scenario where there's only one available action. In this case, we'll always set `function_call` to that action is that expected?

File: front/lib/api/assistant/agent.ts (hunk 4 above)
Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:417]: This allows to use the typs system to ensure we didn't forget to handle an action type. Similar to an assertNever (but here it's "assert action is indeed a generation config"). As explained in the comment above

**Review by flvndvd — APPROVED** [review]
LGTM

**Comment by spolu** [comment]
LGTM \o/

**Pull request merged**